xtext-website/documentation/103_domainmodelnextsteps.md

```diff
@@ -189,6 +189,56 @@ class DomainmodelGenerator extends AbstractGenerator {
 
+## Creating Custom Validation Rules {#tutorial-validation}
+
+One of the main advantages of DSLs is the possibility to statically validate domain-specific constraints. Because this is a common use case, Xtext provides a dedicated hook for this kind of validation rules. In this lesson, we want to ensure that the name of an *Entity* starts with an upper-case letter and that all features have distinct names across the inheritance relationship of an *Entity*.
+
+Locate the class *DomainmodelValidator* in the package *org.example.domainmodel.validation* of the language project. Defining the constraint itself is only a matter of a few lines of code:
+
+```java
+public static final String INVALID_NAME = "invalidName";
+
+@Check
+public void checkNameStartsWithCapital(Entity entity) {
+    if (!Character.isUpperCase(entity.getName().charAt(0))) {
+        warning("Name should start with a capital", 
+            DomainmodelPackage.Literals.TYPE__NAME,
+            INVALID_NAME);
+    }
+}
+```
+
+Any name for the method will do. The important thing is the [Check]({{site.src.xtext_core}}/org.eclipse.xtext/src/org/eclipse/xtext/validation/Check.java) annotation that advises the framework to use the method as a validation rule. If the name starts with a lower case letter, a warning will be attached to the name of the *Entity*.
+
+The second validation rule is straight-forward, too. We traverse the inheritance hierarchy of the *Entity* and look for features with equal names.
+
+```java
+@Check
+public void checkFeatureNameIsUnique(Feature feature) {
+    Entity superEntity = ((Entity) feature.eContainer()).getSuperType();
+    while (superEntity != null) {
+        for (Feature other : superEntity.getFeatures()) {
+            if (Objects.equal(feature.getName(), other.getName())) {
+                error("Feature names have to be unique", DomainmodelPackage.Literals.FEATURE__NAME);
+                return;
+            }
+        }
+        superEntity = superEntity.getSuperType();
+    }
+}
+```
+
+The sibling features that are defined in the same entity are automatically validated by the Xtext framework, thus they do not have to be checked twice. Note that this implementation is not optimal in terms of execution time because the iteration over the features is done for all features of each entity.
+
+You can determine when the `@Check`-annotated methods will be executed with the help of the the [CheckType enum]({{site.src.xtext_core}}/org.eclipse.xtext/src/org/eclipse/xtext/validation/CheckType.java). The default value is `FAST`, i.e. the checks will be executed on editing, saving/building or on request; also available are `NORMAL` (executed on build/save or on request) and `EXPENSIVE` (executed only on request). 
```

Comment:
Typo: `the the`

Reply:
In order to improve the readability, only the word `CheckType` should be marked as a hyperlink instead of `CheckType enum`.